lib/CL/devices/level0/level0-compilation.cc

```diff
@@ -0,0 +1,674 @@
+  });
+}
+
+// ***********************************************************************
+// ***********************************************************************
+
+bool Level0CompilerThread::init() {
+  ze_context_desc_t ContextDescription = {};
+  ContextDescription.stype = ZE_STRUCTURE_TYPE_CONTEXT_DESC;
+  ContextDescription.pNext = nullptr;
+  ContextDescription.flags = 0;
+  ze_result_t ZeRes =
+      zeContextCreate(DriverH, &ContextDescription, &ThreadContextH);
+  if (ZeRes != ZE_RESULT_SUCCESS) {
+    POCL_MSG_ERR("Compiler thread: failed to create L0 Context\n");
+    return false;
+  }
+  Thread = std::thread(&Level0CompilerThread::run, this);
+  return true;
+}
+
+void Level0CompilerThread::run() {
+  bool ShouldExit = false;
+  do {
+    Level0CompilationJobSPtr Job(nullptr);
+    Job = JobQueue->getWorkOrWait(PreferredDeviceH, ShouldExit);
+    if (Job.get() != nullptr) {
+      compileJob(std::move(Job));
+    }
+  } while (ShouldExit == false);
+}
+
+void Level0CompilerThread::compileJob(Level0CompilationJobSPtr Job) {
+  ze_device_handle_t DeviceH = Job->getDevice();
+  Level0ProgramBuild *Build = Job->getBuild();
+  Build->compile(ThreadContextH, DeviceH);
+  Job->signalFinished();
+}
+
+Level0CompilerThread::~Level0CompilerThread() {
+  if (Thread.joinable())
+    Thread.join();
+  if (ThreadContextH) {
+    ze_result_t Res = zeContextDestroy(ThreadContextH);
+    if (Res != ZE_RESULT_SUCCESS) {
+      POCL_MSG_ERR("Compiler thread: failed to destroy L0 Context\n");
+    }
+  }
+}
+
+// ***********************************************************************
+// ***********************************************************************
+
+bool Level0CompilationJobScheduler::init(
+    ze_driver_handle_t H, std::vector<ze_device_handle_t> &DevicesH) {
+
+  JobQueue.reset(new Level0CompilerJobQueue());
+  DriverH = H;
+  unsigned NumThreads = std::thread::hardware_concurrency();
+  unsigned NumDevices = DevicesH.size();
+  assert(NumDevices > 0);
+  for (unsigned i = 0; i < NumThreads; ++i) {
+    ze_device_handle_t PreferredDeviceH = DevicesH[i % NumDevices];
+    CompilerThreads.emplace_back(
+        new Level0CompilerThread(JobQueue.get(), PreferredDeviceH, DriverH));
+  }
+  for (unsigned i = 0; i < NumThreads; ++i) {
+    if (!CompilerThreads[i]->init()) {
+      POCL_MSG_ERR("Failed to initialize CompilerThread %u\n", i);
+      return false;
+    }
+  }
+  return true;
+}
+
+void Level0CompilationJobScheduler::cancelAllJobsFor(Level0Program *Program) {
+  JobQueue->cancelAllJobsFor(Program);
+}
+
+void Level0CompilationJobScheduler::addCompilationJob(
+    Level0CompilationJobSPtr Job) {
+  JobQueue->pushWork(Job);
+}
+
+Level0CompilationJobScheduler::~Level0CompilationJobScheduler() {
+  JobQueue->clearAndExit();
+  CompilerThreads.clear();
+}
+
+// ***********************************************************************
+// ***********************************************************************
+
+/*
```

Comment:
If you want to leave the commented out code, use #if 0 guards and add a comment why it's left there and what's to do to make it usable.

Reply:
ok, i've uncommented both, they are usable but currently unused. might be useful in future.